Integration of WMAS test runner into WPT

louaybassbouss opened the pull request:
Adding [WMAS (Web Media API Snapshot) test runner](https://github.com/cta-wave/WMAS/) for running tests on embedded devices (TVs, Set-Top-Boxes, Streaming devices) in a single window/tab with server-side monitoring and execution. The [WMAS Specification](https://github.com/w3c/webmediaapi/) targets devices that are not able to run the WPT test runner due performance limitations and missing support of multiple windows/tabs (TV Apps run usually in a single window and not able to use window.open() which is used by the WPT Test Runner).   
PS: this [PR](https://github.com/web-platform-tests/rfcs/pull/23) in the [web-platform-tests RFCs](https://github.com/web-platform-tests/rfcs) provides more info about the discussion of the integration of the WMAS Test Runner into WPT.

Additions/Modifications:
- testharnessreport.js: Added logic to communicate with the server
  backend: send results, get URL to next test (only if wpt is started with the command `wpt serve-wave`)
- tools/wptserve/wptserve/wave: Code for WMAS test runner backend written in Python
- wpt code base: Added serve-wave command to load code for backend
  server. Added --report option to activate report generation,
  requires to install wptreport tool.
- Added configuration options to change the servers behaviour
- Documentation inside tools/wptserve/wptserve/wave/docs directory
- Postman API tests inside tools/wptserve/wptserve/wave/test
- Webinterface available under /wave/index.html
- results and generated reports are persisted in /results directory
  (can be configured)
- updated README.md with command `wpt serve-wave`

FritzHeiden committed:
Adding WMAS test runner for running tests on embedded devices in a single
window with serverside monitoring and execution.

Additions/Modifications:
- testharnessreport.js: Added logic to communicate with the server
  backend: send results, get URL to next test
- tools/wave: Code for backend server
- wpt code base: Added serve-wave command to load code for backend
  server. Added --report option to activate report generation,
  requires to install wptreport tool.
- Added configuration options to change the servers behaviour
- Documentation inside tools/wptserve/wptserve/wave/docs directory
- Postman API tests inside tools/wptserve/wptserve/wave/test
- Webinterface available under /wave/index.html
- results and generated reports are persisted in /results directory
  (can be configured)
- updated README.md with command `wpt serve-wave`
- using modified testharnessreport.js only when started using serve-wave

Hexcles commented:
This seems... scary? I never expected this to be 40k lines of code... My browser won't even render the diff page. Are all these changes necessary? Is there a breakdown of this patch? I highly suspect there's a new test suite here as well as new infra.

jgraham commented:
So I've only skimmed this *very* briefly, but some initial thoughts:
* There are a few changes to existing code that I don't understand and which look like they might be reverting some Python 3 fixes (e.g. by changing imports). For the modifications that aren't obviously wave related, can you add some inline comments on the intent of the change (or revert cases that aren't intentional).
* The stuff under `tools/wptserve/wptserve/wave` doesn't look like it's actually modifying the core HTTP server, but creating a new kind of runner that works on top of the server. I suspect that it ought to be in `tools/wave` instead.
* Modifying `testharnessreport.js` shouldn't be necessary; if wave-specific behaviour is required you can just add a route for your specific variant of the file that's installed when `wpt wave-serve` is run. wptrunner also does that.

I don't know if it's possible to split this into multiple PRs, but if there are generic parts that can be pulled out, it would be very helpful. It's very difficult to review such a large changeset.

louaybassbouss commented:
> This seems... scary? I never expected this to be 40k lines of code... My browser won't even render the diff page. Are all these changes necessary? Is there a breakdown of this patch? I highly suspect there's a new test suite here as well as new infra.

Thanks @Hexcles for your feedback. This is almost the work done within the last 2 years in WMAS Test Runner and yes it is big. It adds a new vertical to WPT. We tried to put everything related to the new runner in the separate folder `tools/wptserve/wptserve/wave` to avoid conflicts and make it easier for you to merge. We already excluded everything that is only needed for WMAS and not relevant for WPT.

louaybassbouss commented:
> So I've only skimmed this _very_ briefly, but some initial thoughts:
> 
Thanks @jgraham for you comments, please find my answers below:
>
> * There are a few changes to existing code that I don't understand and which look like they might be reverting some Python 3 fixes (e.g. by changing imports). For the modifications that aren't obviously wave related, can you add some inline comments on the intent of the change (or revert cases that aren't intentional).

Agree we will check and add comments where it is needed. Basically, we tried to avoid changing existing files if there is no need. 

> * The stuff under `tools/wptserve/wptserve/wave` doesn't look like it's actually modifying the core HTTP server, but creating a new kind of runner that works on top of the server. I suspect that it ought to be in `tools/wave` instead.

Exactly this was our intention to not modify the core HTTP Server and add our logic on top of it. We will more the wave related code to `tools/wave` as you recommended.

> * Modifying `testharnessreport.js` shouldn't be necessary; if wave-specific behaviour is required you can just add a route for your specific variant of the file that's installed when `wpt wave-serve` is run. wptrunner also does that.

Our idea behind the new `wpt serve-wave` command is to extend  `wpt serve` and not replace it.  This means, you can still use the WPT test runner via `http://web-platform.test:8000/tools/runner/index.html` even if you use the command `wpt serve-wave`. We can implement you approach and have a WAVE-specific `testharnessreport.js` which is used only if  the server is started via `wpt serve-wave`. 
> 
> I don't know if it's possible to split this into multiple PRs, but if there are generic parts that can be pulled out, it would be helpful. It's very difficult to review such a large changeset.

This PR adds a new Test Runner into WPT and we already skipped everything that is only relevant for WAVE/WMAS. What we submitted is the core the new test runner. Future PRs (bug fixes, adding new features, ...) will be much smaller than this PR.

FritzHeiden commented:
> Modifying testharnessreport.js shouldn't be necessary; if wave-specific behaviour is required you can just add a route for your specific variant of the file that's installed when wpt wave-serve is run. wptrunner also does that.

@jgraham We are currently working on this issue, however we could not find any code that controls what testharnessreport.js is served to requests for `/resources/testharnessreport.js`.
So far we tried changing the path in `tools/wptrunner/wptrunner/environment.py` line 182 without success.

Where does the code responsible for this reside, or where do we implement the desired behaviour?

FritzHeiden commented:
> Where does the code responsible for this reside, or where do we implement the desired behaviour?
Nevermind, we found a solution.

louaybassbouss commented:
@jgraham we made the changes you proposed. Please let us know if you still have questions and how to proceed?

JohnRiv commented on tools/serve/serve.py:
```diff
@@ -902,6 +976,46 @@ def handle_signal(signum, frame):
                 logger.info("Status of %s:\t%s" % (item.name, "running" if item.is_alive() else "not running"))
 
 
+# Set command is_wave and start venv wit necessary dependencies
```
```suggestion
# Set command is_wave and start venv with necessary dependencies
```

JohnRiv commented on tools/wave/docs/rest-api/sessions-api/update.md:
```diff
@@ -0,0 +1,99 @@
+  "reference_tokens": [
+    "bb7aafa0-6a92-11e9-8ec2-04f58dad2e4f",
+    "a50c6db0-6a94-11e9-8d1b-e23fc4555885"
+  ]
```
```suggestion
  ],
```

JohnRiv commented on tools/wave/www/configuration.html:
```diff
@@ -0,0 +1,1117 @@
+//          token: "a50c6db0-6a94-11e9-8d1b-e23fc4555885",
+//          icon: "fab fa-chrome"
+//        }
+//      ];
```
let's remove this commented-out code as well as the commented-out code above it

JohnRiv commented on tools/wave/www/lib/wave-service.js:
```diff
@@ -0,0 +1,864 @@
+  //   if (!token) return;
+  //   const sessionUrl = `/results.html?token=${token}`;
+  //   window.open(sessionUrl, "_blank");
+  // },
```
let's remove all of this commented-out code

JohnRiv commented on tools/wave/www/next.html:
```diff
@@ -0,0 +1,115 @@
+      //     xhr.setRequestHeader(header, headers[header]);
+      //   }
+      //   xhr.send(data);
+      // }
```
remove commented-out code please

JohnRiv commented on tools/wave/www/results.html:
```diff
@@ -0,0 +1,1582 @@
+    <title>Session Results - Web Platform Test</title>
+    <link rel="stylesheet" href="css/bulma-0.7.5/bulma.min.css" />
+    <link rel="stylesheet" href="css/fontawesome-5.7.2.min.css" />
+    <!-- <link rel="stylesheet" href="css/result.css" /> -->
```
remove commented-out code please

JohnRiv commented on tools/wave/www/results.html:
```diff
@@ -0,0 +1,1582 @@
+            //    testFilesCount[api] === testFilesCompleted[api];
+            //  results[api].testFilesCount = testFilesCount[api];
+            //  results[api].testFilesCompleted = testFilesCompleted[api];
+            //}
```
remove commented-out code please

JohnRiv commented on tools/wpt/commands.json:
```diff
@@ -52,8 +52,8 @@
-      "mozdownload",
-      "mozinstall"
+	    "mozdownload",
+	    "mozinstall"
```
undo all the whitespace changes in this file please. We don't want to muddy-up the git history

JohnRiv reviewed (commented):
@louaybassbouss in addition to addressing the merge conflict with `tools/serve/serve.py` that still seems to be there, I've made some minor suggestions inline to help clean this up for the WPT team.

Let's also get rid of the non-minified versions of bulma.css & fontawesome.css (make sure we're not referencing those as well, as we should only reference the minimized ones)

JohnRiv commented:
Thanks, looks like we're free of merge conflicts now.

@jgraham or anyone else, please let us know if you have any additional requests/suggestions in order for this to be merged in. Thank you.

louaybassbouss commented:
Thanks @JohnRiv for your comments all conflicts are resolved.

jgraham reviewed (commented):
I've made a commit at https://github.com/web-platform-tests/wpt/tree/wms_changes that shows how we can factor the serve.py changes so that they're less invasive but use some (new in that commit) well-defined integration points. I think we should make a refactoring like that one.

There's still quite a bit here I haven't looked at in detail.

jgraham commented:
One more comment today whilst I remember: this needs some tests that run when the wpt-infrastructure tests are run in CI. The kind of thing I have in mind is basic integration type tests like ensuring we can start the server and access some wave-specific page. But we should also run any unittests that you write as part of that job.

jgraham commented on tools/wave/testing/test_loader.py:
```diff
@@ -0,0 +1,192 @@
+        self._tests[MANUAL] = {}
+        self._api_titles = api_titles
+
+    def load_tests(self, manifest_file_path):
```
Is this trying to read a `MANIFEST.json` file? You really really want to use the existing libraries for that rather than hand rolling it, because otherwise it will be liable to break whenever the format changes.

jgraham commented on tools/wave/wave_server.py:
```diff
@@ -0,0 +1,103 @@
+
+        configuration = configuration_loader.load(configuration_file_path)
+
+        print(u" done.")
```
Debugging code?

jgraham replied:
Debugging code?

jgraham commented on tools/wave/wave_server.py:
```diff
@@ -0,0 +1,103 @@
+                   application_directory_path=u"",
+                   reports_enabled=False):
+        sys.stdout.write(u"Loading configuration ...")
+        sys.stdout.flush()
```
Why are you writing directly to stdout? Also calling flush is a little odd.

jgraham replied:
Why are you writing directly to stdout? Also calling flush is a little odd.

jgraham reviewed (dismissed):
OK, I've done a pretty cursory review of the Python code, basically looking for worrying code rather than knowing how the whole thing fits together. Overall there are a few high level points:

* Avoid custom rolling code that interacts with other parts of WPT (e..g the manifest). Reusing existing code where possible is one of the best ways to ensure that this doesn't keep breaking.
* Actually ensuring this doesn't keep breaking requires some kind of tests that we can run in CI. I'm happy to help with the CI configuration.
* For third party libraries we need to be careful to ensure that we're distributing the license with the code (and also that the license is acceptable of course).

I think if we can address those issues we'll be in a place to land this.

louaybassbouss commented:
@jgraham thanks for your valuable feedback. FYI we are working on your comments and we are going to provide an update following your recommendation soon.

FritzHeiden committed:
Changes regarding posed requirements in pull request:
- Moved third party frontend library into vendor specific directory
- Added License file for third party frontend library
- Moved wave configuration into wave object inside the configuration object
- Moved rules of tools/wave/.gitignore into tools/.gitignore
- Using context manager when performing file input/output
- Using urlparse.urlunsplit to create urls from individual parts
- Changed comment to express strings as "text type" instead of "str type"
- Moved repeated code for exception handling in API handlers into separate function
- Simplified parsing of urls and query strings
- Removed mutable default parameters
- Using future unicode literals to remove necessity of defining string type for unicode strings
- Using shell=False when executing subprocesses
- Using WPT functions to load tests from MANIFEST.json
- Using proper logger to log messages
- Removed unused ErrorHandler
- Moved wave specific code from serve.py into wave.py
- Changed default uri namespace of wave files and api to "/_wave/"
- Updated rest API documentation

louaybassbouss commented on tools/wave/docs/rest-api/README.md:
```diff
@@ -0,0 +1,54 @@
+
+| Name                                           | Description                                          |
+| ---------------------------------------------- | ---------------------------------------------------- |
+| [`create`](./sessions-api/create.md)           | Creates a new test session.                          |
```
The documentation for the WAVE test runner is written in markdown, which is split up into multiple files that reference each other. In the example you pointed out, we link to the `create.md` file inside the docs directory. Hence the relative path, starting with `./`.

louaybassbouss replied:
Thanks @jgraham again for your feedback. We solved most of the issues you recommended what is still missing are the integration and unit tests. It would be great if you can guide us on this part. As you can also see, the lastet version causes some CI checks to fail. From the logs we can see that this CI test `ci/tc/tests/test_valid.py F.....` fails. Any idea what could be the problem here?

jgraham replied:
I see. Could we make the default `_wave` or something so that it can't clash with any future spec? We have effectively reserved a few words (`tools`, `resources`, etc.) but since communicating that with the spec community is hard it would be good to avoid reserving more (and this case would be particularly tricky since the existing name won't show up in a directory listing.

louaybassbouss replied:
Thanks @jgraham yes we will change it to `_wave`

louaybassbouss replied:
Now default namespace is `_wave`

louaybassbouss commented on tools/serve/serve.py:
```diff
@@ -327,7 +330,9 @@ def __init__(self):
                           ("*", "/tools/*", handlers.ErrorHandler(404)),
-                          ("*", "{spec}/tools/*", handlers.ErrorHandler(404))]
+                          ("*", "{spec}/tools/*", handlers.ErrorHandler(404)),
+                          ("*", "/serve.py", handlers.ErrorHandler(404)),
```
`("*", "/serve.py", handlers.ErrorHandler(404))` removed

louaybassbouss commented:
Thanks @jgraham again for your feedback. We solved most of the issues you recommended what is still missing are the integration and unit tests. It would be great if you can guide us on this part. As you can also see, the lastet version causes some CI checks to fail. From the logs we can see that this CI test `ci/tc/tests/test_valid.py F.....` fails. Any idea what could be the problem here?

jgraham commented:
The failing unittest was a problem on master; I think rebasing would solve it. 

Another high level comment: did you look at the proposed refactor in https://github.com/web-platform-tests/wpt/tree/wms_changes ? I think that would be an improvement for the integration code.

More detailed review later.

louaybassbouss commented:
> The failing unittest was a problem on master; I think rebasing would solve it.

Thanks we will try it.

> Another high level comment: did you look at the proposed refactor in https://github.com/web-platform-tests/wpt/tree/wms_changes ? I think that would be an improvement for the integration code.

We got 404 when we tried to open the link but I now I see that there is a branch `wmas_changes` (correct link is https://github.com/web-platform-tests/wpt/tree/wms_changes). We will check it. 

> 
> More detailed review later.

Thanks

louaybassbouss commented:
@jgraham we integrated your recommended changes in branch [wmas_changes](https://github.com/web-platform-tests/wpt/tree/wmas_changes) (commit [38080eefea578c5e1f9081f0b5ced5ff9a62d260](https://github.com/web-platform-tests/wpt/commit/38080eefea578c5e1f9081f0b5ced5ff9a62d260)) but many integration test failed (local tests were successful). We switched in commit [a7b6ac499a3f3f78d89abb9f350617f1bee01fe2](https://github.com/web-platform-tests/wpt/pull/21323/commits/a7b6ac499a3f3f78d89abb9f350617f1bee01fe2) back to the previous structure and all checks were successful. The namespace is changed now to `_wave`. Any guidance from your side on this?

FritzHeiden commented:
@jgraham We performed a rebase using the current master (d7a573df46aac1ebc338ff993c54a18f2b9ff1a4) and ran tox tests using python 2.7.17, 3.6.10 and 3.8.2 with no errors. However, there are errors related to python 3.8.2 tests in the CI pipeline: test_manifest.py: Data generation is extremely slow
Can you give us some information on how to proceed? Thank you

jgraham commented:
I'd like to solve the problem with the proposed refactor rather than revert it.

Don't worry about the hypothesis failures, that seems to be an unrelated problem.

FritzHeiden commented:
We reintroduced the refactor of putting wave specific logic into a separate file rather than into serve.py. Furthermore, we ran tox tests for several python versions without any errors.

foolip commented:
@FritzHeiden here's the error I see in one Taskcluster job:
```
sudo sh -c ./wpt make-hosts-file >> /etc/hosts
Traceback (most recent call last):
  File "./wpt", line 12, in <module>
    wpt.main()
  File "/home/test/web-platform-tests/tools/wpt/wpt.py", line 167, in main
    rv = script(*args, **kwargs)
  File "/home/test/web-platform-tests/tools/ci/make_hosts_file.py", line 18, in run
    ssl={"type": "none"})
  File "/home/test/web-platform-tests/tools/serve/serve.py", line 831, in build_config
    rv = config_cls()
TypeError: 'str' object is not callable
```

Have you tried to update make_hosts_file.py to pass in the new first argument?

FritzHeiden commented:
@foolip Thanks for your response. I was aware about this potential fix, however, I was not sure if this is the proper solution, as it affects the CI scripts, which I am not familiar with.
I now pushed the proposed change, but there seem to be more problems with the changed parameter list. I am unable to see those errors before pushing it into the PR, as my tox test run seem to pass all tests.
Is there a way to see those errors locally?

FritzHeiden commented:
After fixing the issue it now looks good. There still is a failing check, however, it doesn't look like its due to our changes.

louaybassbouss commented:
Thanks @foolip @jgraham for feedback and @FritzHeiden for fixing the issues. @foolip @jgraham please guide us how to proceed/what you still need from us.

stephenmcgruer commented:
`wpt-chrome-dev-results` failed this time, again network issues, but this change doesn't actually require that to run so I suggest you ignore it @foolip and @jgraham .

foolip commented:
In that case we just need an approving review and then we can force merge. @jgraham, you had requested changes, can you take another look?

foolip commented:
Ping @jgraham

louaybassbouss commented:
@foolip @jgraham any update on this PR?

jgraham commented:
Sorry I've been swamped. If I haven't taken another look by the end of the week please yell at me.

jgraham commented:
I've taken another look, sorry for the delay.

The integration code looks good now, thanks. I'm still not sure how we're going to tell if this gets broken; do we have any tests for this in the CI configuration? I can't see anything but it's also easy to miss in the huge number of changes.

FritzHeiden commented:
@jgraham Thanks for your reply. We somehow did not notice it until now.
We already tried to understand how the CI pipeline works in this project, however, we don't fully understand it. Do you have some resources or tips on how to write tests for it? What config files do we have to modify, where do we put our tests?
So far we understand that you use tox to invoke some pytest commands. I will try to replicate the way you are testing other components using those tools.

stephenmcgruer commented:
@FritzHeiden as you found, for integration/unittests, we rely on tox. We have two different CI systems, Azure Pipelines (for MacOS and Windows, controlled by `.azure-pipelines.yml`) and Taskcluster (for Linux, controlled by a decision task that uses `tools/ci/tc/tasks/test.yml` as its input). Taskcluster is the main flow currently, Azure Pipelines is considered an 'fyi'.

Both of these systems run certain tasks based on what was changed in a given pull request. For example, in `tools/ci/tc/tasks/test.yml` there is:

```
  - tools/ unittests (Python 2):
      use:
        - wpt-base
        - trigger-pr
        - tox-python2
      description: >-
        Unit tests for tools running under Python 2.7, excluding wptrunner
      command: ./tools/ci/ci_tools_unittest.sh
      env:
        HYPOTHESIS_PROFILE: ci
      schedule-if:
        run-job:
          - tools_unittest
```

This isn't the easiest to read, but in short it triggers for pull requests (`trigger-pr`) if the change affects anything in `tools/` (`tools_unittest` is just a named regex, see `tools/ci/jobs.py`). It then runs `./tools/ci/ci_tools_unittest.sh`, which is really a wrapper for tox.

What actually runs there is controlled by the `tox.ini` file in `tools/`, just like any tox-based testing project (if you're familiar).

This is a long way of saying that if running `tox -e py27` in `tools/` currently runs your integration tests*, then they should also run whenever something that touches `tools/` changes, and we should be good. Does that make sense? Is that the case?

* E.g. if I run `tox -e py27` in `tools/` today, I get:

```
...

ci/tc/tests/test_decision.py ..................                                                  [  2%]
ci/tc/tests/test_taskgraph.py .........                                                          [  3%]
ci/tc/tests/test_valid.py ......                                                                 [  4%]
ci/tests/test_jobs.py ............                                                               [  6%]
ci/tests/test_pr_preview.py ..................                                                   [  9%]
gitignore/tests/test_gitignore.py ...................................................            [ 16%]
lint/tests/test_file_lints.py .................................................................. [ 26%]
...................................                                                              [ 31%]
lint/tests/test_lint.py .........................................                                [ 37%]
lint/tests/test_path_lints.py ...............................                                    [ 42%]
manifest/tests/test_XMLParser.py .........                                                       [ 43%]
manifest/tests/test_item.py .............................                                        [ 47%]
manifest/tests/test_manifest.py .........                                                        [ 49%]
manifest/tests/test_sourcefile.py .............................................................. [ 58%]
..................................................................                               [ 68%]
manifest/tests/test_utils.py .                                                                   [ 68%]
serve/test_functional.py .                                                                       [ 68%]
serve/test_serve.py .s.............                                                              [ 70%]
wptserve/tests/test_config.py .......xx.....................................                     [ 77%]
wptserve/tests/test_replacement_tokenizer.py ..............                                      [ 79%]
... etc
```

louaybassbouss commented:
Thanks @stephenmcgruer for the explanation we will check and come back to you if we have questions.

FritzHeiden committed:
Implemented CI test for WAVE test runner
- made wave code compatible with python 3
- implemented test to start the wave server and ping an API endpoint
- moved adding --report argument to wave.py
- added WAVE test runner hint to test error messages
- removed double declaration of build_config function in serve.py

FritzHeiden commented:
I added `tools/wave/tests/test_wave.py` which spins up the WPT server with `serve-wave` argument and checks if one of our API endpoints returns status code 200.

Furthermore, when I run tox locally I get an `ImportMismatchError` for some python 2.7 call. This also occurs on the current master branch.

stephenmcgruer commented on tools/serve/serve.py:
```diff
@@ -925,6 +886,43 @@ def _get_paths(self, data):
         return rv
 
 
+def build_config(override_path=None, config_cls=ConfigBuilder, **kwargs):
```
Moving the entire function makes it very difficult to see what has changed here. Please don't do that - if you're making changes to this function leave it on line 783 unless there's a specific reason not to.

FritzHeiden replied:
Due to some merging error in the previous commit, there were two build_config function. Apparently, I removed the wrong one

FritzHeiden replied:
I see, there was a reason the function was moved: We moved the wave logic from serve.py to wave.py and as @jgraham suggested, we pass the config builder class to the function, so we can have a different one defined in the wave.py. For this to not brake other code we added the default value of the local ConfigBuilder to the paramter, which is only accessible after its declaration. If you have another suggestion for this, please let me know.

stephenmcgruer replied:
Ack, that sounds like a specific reason to keep this to me :). Thanks for the answer!

stephenmcgruer commented:
> I added tools/wave/tests/test_wave.py which spins up the WPT server with serve-wave argument and checks if one of our API endpoints returns status code 200.

Awesome, we should be able to see in https://community-tc.services.mozilla.com/tasks/aZjv179XScCjX-3oCM59jQ if that worked or not.

> Furthermore, when I run tox locally I get an ImportMismatchError for some python 2.7 call. This also occurs on the current master branch.

I'm not totally surprised; running tox directly is a bit of a hack that mostly works if you happen to have the right setup. Things are only really meant to work when run in the docker container, which  the CI system runs in. I'm just lazy and often run tox locally :)

stephenmcgruer commented:
> Awesome, we should be able to see in https://community-tc.services.mozilla.com/tasks/aZjv179XScCjX-3oCM59jQ if that worked or not.

Looks like this failed - [logs](https://community-tc.services.mozilla.com/tasks/aZjv179XScCjX-3oCM59jQ/runs/0/logs/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FaZjv179XScCjX-3oCM59jQ%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log).

```
...
wave/tests/test_wave.py F                                                [ 70%]
...
=================================== FAILURES ===================================
__________________________________ test_serve __________________________________
    def test_serve():
        if is_port_8080_in_use():
            assert False, "WAVE Test Runner failed: Port 8080 already in use."
    
        p = subprocess.Popen([os.path.join(wpt.localpaths.repo_root, "wpt"),
            "serve-wave",
            "--config",
            os.path.join(wpt.localpaths.repo_root, "tools/wave/tests/config.json")],
            preexec_fn=os.setsid)
    
        start = time.time()
        try:
            while True:
                if p.poll() is not None:
                    assert False, "WAVE Test Runner failed: Server not running."
                if time.time() - start > 60:
>                   assert False, "WAVE Test Runner failed: Server did not start responding within 60s."
E                   AssertionError: WAVE Test Runner failed: Server did not start responding within 60s.
E                   assert False
wave/tests/test_wave.py:44: AssertionError
----------------------------- Captured stdout call -----------------------------
URLError
URLError
URLError
URLError
URLError
URLError
URLError
URLError
URLError
...
```

FritzHeiden committed:
Changed conditions for wave CI test

- increased startup timeout to 6 minutes
- updated name in README to WAVE Test Runner

FritzHeiden commented:
I didn't factor in the time it takes to generate the MANIFEST.json, so I increased the timeout for the server startup in the test.

FritzHeiden committed:
Changed test implementation to be compatible with windows

- removed preexec_fn from Popen call, which is only supported for unix
  systems

FritzHeiden commented:
I don't know how to resolve the problems with the wave test in the CI pipeline. It seems like there already is an instance of the WPT running, which I already tried to circumvent by running it on different ports, however this doesn't seem to work. Furthermore, on windows there is this error that I don't understand what causes it: `OSError: [WinError 193] %1 is not a valid Win32 application`

My approach to implement the wave test is, that I saw this one test, `test_serve` in `tools/wpt/tests/test_wpt.py`, which also starts the WPT runner and checks if it returns a status of 200. I used this function for the wave test by modifying it accordingly. I figured there are some APIs used in the test, that are not available for windows, so I removed them.

When I run the test locally using tox, everything seems to work just fine. However, as this was pointed out to me, my local environment does not guarantee a successful test run in the pipeline, so this is not a reliable indicator.

I noticed the test I used the code from is only run on macOS in the pipeline and it does not seem to have the same issues the wave test has in regards to the ports being already in use.

Could you provide more information on how to properly write this test? Thank you

stephenmcgruer commented:
> I noticed the test I used the code from is only run on macOS in the pipeline and it does not seem to have the same issues the wave test has in regards to the ports being already in use.

It should be run on Linux too, it only has a pyskip for Windows as far as I can see. It does scare me that every test there has a pyskip for port 8000 being already in use, that seems like a bit of a yikes...

> It seems like there already is an instance of the WPT running

I ran it on its own in https://github.com/web-platform-tests/wpt/pull/23955, and it still failed. [Logs](https://community-tc.services.mozilla.com/tasks/Vsk6WqmFTuOn8_EOGFXwew/runs/0/logs/live/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FVsk6WqmFTuOn8_EOGFXwew%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log). I started adding logging, and it almost looks like it fails in loading the manifest? I would suggest copying my commits from that PR, to only run the wave test, and then adding more logging to investigate what is going on.

stephenmcgruer commented:
A few things I've noticed as I continue to prod at this:

1. On the CI, WPT manifest generation takes 160 seconds with parallel mode **on** (it was originally off in the wave server, not sure why). cc @jgraham @gsnedders that seems like a long time to me, is that expected? Should some process be creating the manifest ahead of tests that may want to use it? (Note that these tests currently run as part of `tools/ unittests` aka `tools/ci/ci_tools_unittest.sh`).
1. It looks like maybe there's a hosts failure; once I got past the manifest generation and hacked some logging in place I saw:
```
CRITICAL:web-platform-tests:Failed to start HTTP server on port 37001; is something already using that port?
CRITICAL:web-platform-tests:Please ensure all the necessary WPT subdomains are mapped to a loopback device in /etc/hosts.
See https://web-platform-tests.org/running-tests/from-local-system.html#system-setup for instructions.
CRITICAL:web-platform-tests:Failed to connect to test server on http://web-platform.test:37001/. Please ensure all the necessary WPT subdomains are mapped to a loopback device in /etc/hosts.
See https://web-platform-tests.org/running-tests/from-local-system.html#system-setup for instructions.
```
I have a version that hacks in a call to `wpt make-hosts-file | sudo tee -a /etc/hosts` running the CI now, we'll see if that helps.

stephenmcgruer commented:
The lack of hosts file appears to have been the other issue; [the latest run](https://community-tc.services.mozilla.com/tasks/NkhGRgtyTRydu3p_RkCISA/runs/0/logs/live/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FNkhGRgtyTRydu3p_RkCISA%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log) looks like its properly passing.

I will take a look tomorrow at how these tests are setup in general, with an eye to finding out:

1. If anything does or should create a manifest for the tests,
1. If anything does or should setup the hosts file, and
1. If this test should be in the same category as `tools/wpt tests` rather than `tools/ unittests`.

louaybassbouss commented:
Thanks a lot @stephenmcgruer for quick reply and your investigations. We will then wait for your guidance on the next steps if you still input from us.

stephenmcgruer commented:
Ok, so there is only get a hosts file setup if the task has `hosts: true` specified for it. This isn't specified for `tools/ unittests` but is for `tools/wpt/ tests` (which the descriptions back; the former is for unittests, the later for 'Integration tests for wpt commands').

So one thing I think we want to do here is make sure `tools/wave/tests/test_wave.py` is run as a WPT integration test. @jgraham do you prefer this test being moved to `tools/wpt/tests`, or `tools/wave/tests` being added to the things that `tools/wpt/ tests` runs, or a third option?

jgraham commented:
>  @jgraham do you prefer this test being moved to tools/wpt/tests, or tools/wave/tests being added to the things that tools/wpt/ tests runs, or a third option?

Hmm. So the setup has a degeneracy between unittest vs integration test and tools/ vs tools/wpt.

Maybe in the long term the best way to fix this would be to have a single tox.ini specifying all the test types and using different environments to group them into unittest/integration-test, by Python version, etc.

In the meantime, maybe it's best to rename the test tasks to "Python Unittests" and "Python Integration tests" and put the wave tests as an additional command in the latter (which currently correspond to the `tools/wpt tests`).

stephenmcgruer committed:
Add tools/wave to the integration tests

stephenmcgruer commented:
@louaybassbouss - ok, I think we've got things in a place where they should work. You need to rebase this PR onto master, and then add the changes I did in https://github.com/web-platform-tests/wpt/pull/23955/commits/9d5fd56bf4d956e90f4afeabd45763b2f094a96f . That moves the wave test to the integration test setup, which should run it successfully.

Note that tools/wave/tox.ini may be over-complicated and we may be able to simplify it in review, I just grabbed a copy of another tox file to be quick :)

louaybassbouss commented:
> @louaybassbouss - ok, I think we've got things in a place where they should work. You need to rebase this PR onto master, and then add the changes I did in [9d5fd56](https://github.com/web-platform-tests/wpt/commit/9d5fd56bf4d956e90f4afeabd45763b2f094a96f) . That moves the wave test to the integration test setup, which should run it successfully.

Great thank @stephenmcgruer we will let you know when we are done

FritzHeiden committed:
Changed implementation of test_wave.py integration test

- Changed used ports to defaults
- Reverted changes made to make test run on windows
- Skipping test when port is used instead of failing the test

louaybassbouss commented:
@stephenmcgruer we did the recommended changes and now all check have passed :)

stephenmcgruer commented:
> @stephenmcgruer we did the recommended changes and now all check have passed :)

Yay! I can confirm that the test appears to be running + passing for the integration tests target. I will review the related changes later today.

Ping @jgraham who may also want to review the new `tox.ini` file at least (not sure)

stephenmcgruer commented on README.md:
```diff
@@ -69,6 +69,8 @@ commands are:
 * `wpt install` - For installing the latest release of a browser or
   webdriver server on the local machine.
+* `wpt serve-wave` - For starting the wpt http server and the WAVE test runner.
+More details on how to use the WAVE test runner see the [documentation](./tools/wave/docs/usage/usage.md).
```
```suggestion
For more details on how to use the WAVE test runner see the [documentation](./tools/wave/docs/usage/usage.md).
```

stephenmcgruer reviewed (approved):
LGTM. Please give @jgraham 24 hours to give any final comments from their side, otherwise after 24 hours I will merge this.

louaybassbouss committed:
Update README.md

Minor update in the `wpt serve-wave` command description

Co-authored-by: Stephen McGruer <stephen.mcgruer@gmail.com>

louaybassbouss commented:
> LGTM. Please give @jgraham 24 hours to give any final comments from their side, otherwise after 24 hours I will merge this.

Thanks a lot @stephenmcgruer
